### PR TITLE
applications: nrf5340_audio: Audio module improvements

### PR DESCRIPTION
**Pull request by gWacey**
Corrections from use, additional testing and fake function updates.

### CODE REVIEW

File: include/audio_module/audio_module.h [hunk 1]
```diff
@@ -23,6 +23,19 @@ extern "C" {
 
 #include "audio_defines.h"
 
+/**
+ * @brief Helper macro to configure the modules parameters.
+ */
+#define AUDIO_MODULE_PARAMETERS(p, dest, sck, sck_size, pri, fifo_rx, fifo_tx, slab, slab_size)    \
```
Comment (koffes) [inline include/audio_module/audio_module.h:29]:
```suggestion
#define AUDIO_MODULE_PARAMETERS(p, dest, stk, stk_size, pri, fifo_rx, fifo_tx, slab, slab_size)    \
```

Comment (koffes) [reply include/audio_module/audio_module.h:29]: SCK is often used for serial clocks

File: subsys/audio_module/audio_module.c [hunk 2]
```diff
@@ -185,14 +187,15 @@ static int data_tx(struct audio_module_handle *tx_handle, struct audio_module_ha
 		}
 
 		/* Copy. The audio data itself will remain in its original location. */
-		memcpy(&data_msg_rx->audio_data, audio_data, sizeof(struct audio_data));
+		memcpy((uint8_t *)&(data_msg_rx->audio_data), audio_data,
```
Comment (koffes) [inline subsys/audio_module/audio_module.c:190]: Why is this casting required now?

File: subsys/audio_module/audio_module.c [hunk 3]
```diff
@@ -562,7 +565,7 @@ int audio_module_open(struct audio_module_parameters const *const parameters,
 	/* Clear handle to known state. */
 	memset(handle, 0, sizeof(struct audio_module_handle));
 
-	/* Allocate the context memory. */
+	/* Store poimter to the context. */
```
Comment (koffes) [inline subsys/audio_module/audio_module.c:568]:
```suggestion
	/* Store pointer to the context. */
```

File: subsys/audio_module/audio_module.c [hunk 4]
```diff
@@ -1117,41 +1137,43 @@ int audio_module_data_tx_rx(struct audio_module_handle *handle_tx,
...
 			ret);
 		return ret;
 	}
 
-	if (msg_rx->audio_data.data == NULL || msg_rx->audio_data.data_size == 0) {
-		LOG_ERR("Data output buffer too small for received buffer from module %s "
-			"(%d)",
-			handle_rx->name, msg_rx->audio_data.data_size);
-		ret = -EINVAL;
+	if (audio_data_rx->data_size != 0) {
```
Comment (koffes) [inline subsys/audio_module/audio_module.c:1154]: This new section looks good, but there is duplicated code with the section above. Please factor out.

File: tests/subsys/audio_module/prj.conf [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2023 Nordic Semiconductor ASA
+# Copyright (c) 2024 Nordic Semiconductor ASA
```
Comment (koffes) [inline tests/subsys/audio_module/prj.conf:2]:
```suggestion
# Copyright (c) 2023 Nordic Semiconductor ASA
```
Keep the earliest year.

File: tests/subsys/audio_module/src/audio_module_test_common.c [hunk 6]
```diff
@@ -8,7 +8,8 @@
 #include <zephyr/fff.h>
 
 #include "audio_module_test_common.h"
-#include "audio_module/audio_module.h"
+#include "audio_module.h"
+#include "fakes.h"
```
Comment (koffes) [inline tests/subsys/audio_module/src/audio_module_test_common.c:12]: "Fakes" is quite generic. Should this have a more specific name?

File: tests/subsys/audio_module/src/audio_module_test_common.c [hunk 7]
```diff
@@ -19,6 +20,7 @@ void test_context_set(struct mod_context *ctx, struct mod_config const *const co
 	memcpy(&ctx->test_string, TEST_STRING, sizeof(TEST_STRING));
 	ctx->test_uint32 = TEST_UINT32;
 	memcpy(&ctx->config, config, sizeof(struct mod_config));
+	reset_fake_fifo_counter();
```
Comment (koffes) [inline tests/subsys/audio_module/src/audio_module_test_common.c:23]: Verb preferred last: 
```suggestion
	fake_fifo_counter_reset();
```

File: tests/subsys/audio_module/src/audio_module_test_common.c [hunk 8]
```diff
@@ -96,7 +98,10 @@ int test_data_process_function(struct audio_module_handle_private *handle,
 	ARG_UNUSED(handle);
 
 	memcpy(audio_data_tx, audio_data_rx, sizeof(struct audio_data));
-	memcpy(audio_data_tx->data, audio_data_rx->data, audio_data_rx->data_size);
+
+	if (audio_data_rx->data != NULL && audio_data_tx->data != NULL) {
+		memcpy(audio_data_tx->data, audio_data_rx->data, audio_data_rx->data_size);
+	}
```
Comment (koffes) [inline tests/subsys/audio_module/src/audio_module_test_common.c:104]: Should there be a warn/fail if this is not true?

Comment (koffes) [reply tests/subsys/audio_module/src/audio_module_test_common.c:104]: It seems strange to write the `data_size` on the line below if there are NULL pointers present.

File: tests/subsys/audio_module/src/fakes.c [hunk 9]
```diff
@@ -28,26 +28,19 @@ struct test_msg_fifo_queue {
...
 	size_t tail;
 	size_t size;
-	size_t locked;
+	struct k_sem sem;
 
 	void **data[FAKE_FIFO_MSG_QUEUE_SIZE];
 };
 
+/* FIFO "slab" */
+/* FIFO "slab" */
```
Comment (koffes) [inline tests/subsys/audio_module/src/fakes.c:37]: Double double comments? ;)

File: tests/subsys/audio_module/src/functional_test.c [hunk 10]
```diff
@@ -948,22 +947,39 @@ ZTEST(suite_audio_module_functional, test_data_tx_fnct)
...
+	zassert_equal(ret, 0, "Data TX function did not return successfully: ret %d", ret);
+
+	ret = data_fifo_pointer_last_filled_get(&fifo_rx, (void **)&msg_rx, &size, K_NO_WAIT);
+	zassert_equal(ret, 0, "Data TX function did not return 0: ret %d", 0, ret);
+	zassert_mem_equal(msg_rx->audio_data.data, test_data, TEST_MOD_DATA_SIZE,
+			  "Failed Data TX, data differs");
 	zassert_equal(msg_rx->audio_data.data_size, TEST_MOD_DATA_SIZE,
 		      "Failed Data TX function, data sizes differs");
-	zassert_equal(data_fifo_pointer_first_vacant_get_fake.call_count, 1,
+	zassert_equal(data_fifo_pointer_first_vacant_get_fake.call_count, 3,
```
Comment (koffes) [inline tests/subsys/audio_module/src/functional_test.c:979]: Why `3` here? Can it be defined?

**Review by koffes — CHANGES_REQUESTED** [review]

File: tests/subsys/audio_module/src/functional_test.c [hunk 11]
```diff
@@ -253,21 +253,15 @@ static void test_close(const struct audio_module_functions *test_fnct,
 	data_fifo_empty_fake.custom_fake = fake_data_fifo_empty__succeeds;
 
 	if (fifo_rx_set) {
-		data_fifo_deinit(&mod_fifo_rx);
-		data_fifo_init(&mod_fifo_rx);
-		handle.thread.msg_rx = &mod_fifo_rx;
-
-		empty_call_count++;
+		data_fifo_init(&fifo_rx);
```
Comment (andvib) [inline tests/subsys/audio_module/src/functional_test.c:256]: Check return values (for all calls to `data_fifo_init` in this file)

**Review by andvib — CHANGES_REQUESTED** [review]
Looks good! Only minor comments - mostly some missing return checks

**Review by koffes — CHANGES_REQUESTED** [review]
I re-requested a review from myself too soon :)

File: subsys/audio_module/audio_module.c (hunk 2 above)
Comment (gWacey) [reply subsys/audio_module/audio_module.c]: Err don't know will remove.

File: subsys/audio_module/audio_module.c (hunk 4 above)
Comment (gWacey) [reply subsys/audio_module/audio_module.c:1159]: Can we discuss?

File: tests/subsys/audio_module/src/audio_module_test_common.c (hunk 6 above)
Comment (gWacey) [reply tests/subsys/audio_module/src/audio_module_test_common.c]: Renamed audio_module_test_fakes.[c,h]

File: subsys/audio_module/CMakeLists.txt [hunk 12]
```diff
@@ -4,4 +4,7 @@
 # SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 #
 
-target_sources(app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/audio_module.c)
+target_sources(app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/audio_module.c
+)
```
[the remark this answers is not in the document]
Comment (gWacey) [reply subsys/audio_module/CMakeLists.txt]: Oops yes!

**Review by andvib — APPROVED** [review]

File: tests/subsys/audio_module/src/audio_module_test_common.c (hunk 8 above)
Comment (koffes) [reply tests/subsys/audio_module/src/audio_module_test_common.c:108]: Discuss. I need to understand the context here a bit better.

**Review by koffes — COMMENTED** [review]
Looks nice :) A couple of topics to discuss.

**Review by koffes — APPROVED** [review]

**Review by andvib — APPROVED** [review]

**Pull request merged**